### DISCUSSION TITLE
.fmf formatting

### DISCUSSION THREAD

**Discussion by martinhoyer** [teemtee/tmt#2164] [Ideas]
[Identify](https://github.com/pre-commit/identify) is a python file identification library. It is being used by pre-commit.  
I would like to add .fmf to the [list of extensions](https://github.com/pre-commit/identify/blob/main/identify/extensions.py) there, but have a couple of questions:  

 - @psss fmf is your baby, can you see any issues with adding it?
 - fmf being based on yaml and being parsed by ruamel.yaml - wouldn't it make sense to add yaml? i.e.:
```
 EXTENSIONS = {
+    'fmf': {'text', 'fmf', 'yaml'},
```
as opposed to just
```
+    'fmf': {'text', 'fmf'},
```
Adding it would mean any pre-commit hooks that are being run against yaml files would also run against .fmf files without explicitly configuring it (which might sometimes not even be possible ...yamlfix :eyes:)

fyi `.fmf` directory is correctly identified.
```
$ identify-cli .fmf
["directory"]
$ identify-cli plans/iscsi/tier0.fmf 
["file", "fmf", "non-executable", "text", "yaml"]
```
I'm not sure where else is identify lib being used, but in pre-commit, this would allow running checks against .fmf files much easier by being able to use `types` and `types_or` filters. 
Also, it would allow to make tmt/.pre-commit-hooks.yaml a bit cleaner.

**Comment by psss** [comment]
Thanks for the idea. Sounds good to me. Yes, `fmf` files will alwas have to be `yaml`, so I guess it makes sense to identify them as such. And if this makes it easier for `pre-commit` to recognize them, the better!

**Comment by martinhoyer** [comment]
oh well, nevermind https://github.com/pre-commit/identify/pull/395
There are indeed other uses for .fmf extension. Found this for example: https://github.com/Tonsil/flipper-music-files

As for formatting, linting .fmf files, I've settled with [yamlfix](https://lyz-code.github.io/yamlfix):
pre-commit-config.yaml:
```
  - repo: https://github.com/lyz-code/yamlfix/
    rev: 1.11.0
    hooks:
      - id: yamlfix
        types:
          - text
        files: \.fmf$
```
pyproject.toml:
```
[tool.yamlfix]
explicit_start = "false"
line_length = 100
preserve_quotes = "true"
sequence_style = "block_style" 
```
which leaves yaml/yml files alone and only checks/formats .fmf files. 

@psss does it make sense to have some tmt-recommended formatting standard, or is it "anything yaml parsable is good"? :)

**Reply by psss** [reply]
I've quickly checked `yamlfix` and it sounds good and seems to be nicely configurable. Would you create a pull request with the config above so that we can experiment with it? Let's see how many changes it will propose :)

> @psss does it make sense to have some tmt-recommended formatting standard, or is it "anything yaml parsable is good"? :)

I guess there are no special rules or recommendations for `fmf` files nor `tmt` config. The only special thing about `fmf` is the virtual hierarchy, that is keys starting with a slash which are introducing a new hierarchy level. For syntax hightlighting that could be nice to differentiate, but otherwise it's all just `yaml` ;-)

**Reply by martinhoyer** [reply]
So I've been playing with yamlfix and yamlfmt today. 
There is one issue with both - I have not found a way to tell them to let entries like Description start from new line. This result in for example:
from:
```
summary:
    Support for multiple configurations
description:
    All steps support multiple configurations. In this way it is
    possible for example to discover tests from multiple repos.
    The 'execute' steps supports only a single configuration.
```
to:
```
summary: Support for multiple configurations
description: All steps support multiple configurations. In this
    way it is possible for example to discover tests from multiple
    repos. The 'execute' steps supports only a single configuration.
```
which is ok somewhere, but isn't as visually nice as it was before reformatting in most cases.

These are the differences in reformatted fmf files between the two tools:
https://github.com/martinhoyer/tmt/compare/65cff1357e470963b0bac6fee5383849c9bfc98f..33c08b293c32cf01f3022bacb1050ae7699d3dcf

 - Both needs to be told to preserve quotes, otherwise will remove needed quotes, which breaks ruamel parsing.
 - Line length of 62 is needed if blocks of text are to be under 72 characters wide :shrug: . 
 - yamlfix whiteline settings doesn't work for all objects; while yamlfmt does not care.
 - yamlfix uses ruyaml, yamlfix ruamel.yaml

That said, @psss do you still want pull request for experimenting?

**Reply by psss** [reply]
Thanks for trying both and providing the nice summary. Hm, so I see that `yamlfix` is removing empty lines which are sometimes used as a visual separator. Can this be turned off? The multiline reformatting is not very nice either, but we could possibly live with that. I've also noticed that `true` is changed to `no`. I believe the `yes` and `no` values are discouraged by the latest `yaml` specification. All in all, I'm not completely convinced about the benefit. @happz, thoughts?

**Reply by martinhoyer** [reply]
> true is changed to no

Actually the yamlfmt changed `no` to `false` see https://github.com/martinhoyer/tmt/blob/main/tests/unit/id/defined/main.fmf. I don't see a option to make it not do that(unless I'm missing something).
yamlfix can be told to do it as well ('True' -> true, 'no' -> 'false')

> yamlfix is removing empty lines which are sometimes used as a visual separator. Can this be turned off?

Not working for me, but there is this parameter https://lyz-code.github.io/yamlfix/#whitelines-adjusting
Maybe can be fixed upstream. The maintainer is responsive.

> I'm not completely convinced about the benefit

+1 in the current state. I would not add it as a pre-commit check and only focus on yaml validation, not formatting.  
That said, it wouldn't hurt to have standardized yamlfix parameters in pyproject.toml, as it doesn't introduce any new file and people can opt-in to use it as a reference.

**Reply by happz** [reply]
+1 for avoiding formatting. Validation sounds like a useful addition, I'd leave the formatting to project owners - they can employ their own yamlformatwhatever & custom rules.

**Reply by psss** [reply]
Ack, let's go with yaml validation for now.

**Reply by martinhoyer** [reply]
> I'd leave the formatting to project owners

Sure, but it is still nice to have a reference formatting. For example for people like me it's quite confusing when there are inconsistencies, regardless whether the ruamel says the syntax is valid or not.
I mean, even fmf files in tmt repo are not consistently formatted :)

In the end, it might make things ever so slightly harder for newcomers that just want to use tmt, not learn yaml.


FYI, looking at the [yaml spec](https://yaml.org/spec/1.2.2/#rule-c-folded), what should actually be used for the multi-line text blocks is `>`. So in the example above, it would be:

```
summary: >
    Support for multiple configurations
description: >
    All steps support multiple configurations. In this way it is
    possible for example to discover tests from multiple repos.
    The 'execute' steps supports only a single configuration.
```
(though I would not use it for single lines, like 'summary' personally)

**Reply by happz** [reply]
> > I'd leave the formatting to project owners
> 
> Sure, but it is still nice to have a reference formatting. For example for people like me it's quite confusing when there are inconsistencies, regardless whether the ruamel says the syntax is valid or not. I mean, even fmf files in tmt repo are not consistently formatted :)

I agree, it's nice & important to have files in a project formatted in a consistent way. I'm not so sure about giving it a fancy label like "reference formatting" though - I always felt this is very project- and team-specific, some teams rule with an iron fist & use nothing but 2 spaces for YAML indentation, while other teams swear by 4 spaces, and at least one of these groups would immediately break the rules of our reference formatting. I for one see no benefit in such an offering :) I'm all for keeping tmt repo formatted consistently, absolutely, including examples in docs and docstrings if at all possible, but I for one wouldn't bother recommending it to others in such an official-ish way.

> In the end, it might make things ever so slightly harder for newcomers that just want to use tmt, not learn yaml.

I believe the consistent formatting in repo would be a good example, worth following, making things much easier for others to adopt. I'd even say it might be enough to help newcomers, without labeling it with a "recommended formatting" or "best practices" stamp.

> FYI, looking at the [yaml spec](https://yaml.org/spec/1.2.2/#rule-c-folded), what should actually be used for the multi-line text blocks is `>`. So in the example above, it would be:
> 
> ```
> summary: >
>     Support for multiple configurations
> description: >
>     All steps support multiple configurations. In this way it is
>     possible for example to discover tests from multiple repos.
>     The 'execute' steps supports only a single configuration.
> ```
> 
> (though I would not use it for single lines, like 'summary' personally)

The `description` I would consider a question for validation to resolve, as picking the right multiline form in YAML can be tricky. `summary` would be formatting in my books, I wouldn't use a multiline string for this either, but some teams might. As long as it'd be consistent in the repo, I'd be fine with pretty much anything. Well, except Black :)

**Reply by martinhoyer** [reply]
@happz 
> I'm not so sure about giving it a fancy label like "reference formatting" though

> I'd even say it might be enough to help newcomers, without labeling it with a "recommended formatting" or "best practices" stamp.  

Indeed. When I wrote "reference", I meant what's in the existing documentation, examples dir. Together with yamlfix config in pyproject.toml, which wouldn't need to be "advertised" as such.
